adr/ADR-42.md

```diff
@@ -0,0 +1,166 @@
+# Pull Consumer Priority Groups
+
+| Metadata | Value             |
+|----------|-------------------|
+| Date     | 2024-05-14        |
+| Author   | @ripienaar        |
+| Status   | Approved          |
+| Tags     | jetstream, server |
+
+
+| Revision | Date       | Author     | Info           |
+|----------|------------|------------|----------------|
+| 1        | 2024-05-14 | @ripienaar | Initial design |
+
+## Context and Problem Statement
+
+We have a class of feature requests that all come down to adding behaviours on a consumer that affects delivery to groups 
+of clients who interact with the consumer.
+
+Some examples:
+
+ * A single client should receive all messages from a consumer and if it fails another should take over
+ * Groups of related clients should be organised such that certain clients only receive messages when various conditions are met
+ * Groups of clients should access a Consumer but some should have higher priority than others for receiving messages
+ * A consumer should deliver messages to clients in a specific grouping such that related messages all go to the same client group
+
+The proposed feature here address some of these needs while paving the way for future needs to be addressed within this
+framework and under this umbrella feature called `Pull Consumer Groups`.
+
+The current focus is around providing building blocks that can be used to solve higher order problems client-side.
+
+Related prior work:
+
+ * Proposed [Consumer Groups](https://github.com/nats-io/nats-architecture-and-design/pull/36)
+ * Proposed [Partitioned consumer groups and exclusive consumer](https://github.com/nats-io/nats-architecture-and-design/pull/263)
+ * Proposed [Consumer Owner ID](https://github.com/nats-io/nats-server/pull/5157)
+
+## General Overview
+
+We introduce 2 settings on `ConsumerConfig` that activates these related features:
+
+```go
+{
+	PriorityGroups: ["<group>", ...],
+	PriorityPolicy: "<policy>"
+}
+```
+
+The presence of the `PriorityGroups` setting activates the set of features documented here with a list of names supported
```

Comment:
After working with the codebase, I came to conclusion that making `PriorityGroups` an activator makes it really unintuitive from client perspective and not very beneficial for future proofing. 
Why client needs to set `PriorityGroups`, which does nothing right now, if they can just set `PriorityPolicy` which is what makes sense to actually check in the code?

I absolutely support the idea of thinking forward with the how we will expand the feature set, I'm just not convinced this is the way to go for above reason and few others:
1. Are we confident that each `PriorityGroup` cannot have it's own `PriorityPolicy`? If yes, `PriorityGroups` should be rather `PriorityGroups: []PriorityGroup`, which would contain group name, policy, etc. Otherwise, we would probably anyway introduce changes that are hard to not make them breaking, or easy to work with.
2. Handlin many priority groups requires a pretty heavy refactor of how we fetch messages. That can impact the API too in a ways we have not yet discovered.

I'm a bit afraid of having confusing API saying "if you want to use `PriorityPolicy`, you need to also specify `PriorityGroups`. It's a slice, but you can add only one member. And it does not matter now."

On the other hand, picking default name for priority groups could work (if `PriorityPolicy is set, and `PriorityGroups` are not, set a a default `PriorityGroups`), but that again will be a usecase for pedantic mode, and can become surprising when we will start actually support `PriorityGroups`.

**I think we should either make `PriorityGroups: []PriorityGroup` straigtht away, and say that only one can be passed as a compromise between future proofing and good ux, or remove the requirement.**

Reply:
Group name is used in pinned clients.  Using it now on day one even though a bit odd being just one means old clients will remain forward compatible once we support more or add meaning to those group names.  They will already have a group name set and continue to work.

On the other hand if we drop the group name requirement and only introduce it later we get a cliff where all older clients just start erroring.  Perhaps thats a good forcing function but I suspect what we'd end up doing in that case is further complicate the server code to try and handle the case of no group and multiple groups - at that point not having set it today becomes the worst possible outcome.

Reply:
Sure, that is right, IF we are 100% sure that we will not want PriorityGroups with different settings each - so `PriorityGroups: []string` vs `PriorityGroups: []PriorityGroup`.

I dont have strong opinion here - I agree with you. But I think that if we want to introduce groups as mandatory public API, we need to think and be confident which of above cases will be future proof.

Reply:
We can certainly explore if the list of policy is the right approach - I agree a map might be better long run.  

I just know what we'll do when the cliff comes - we'll complicate the server to deal with our earlier minimal API design rather than force upgrades so I think the effort to explore some possible futures a bit more would be worth it.

Reply:
We discussed this and thought making groups maps would get us into a complexity trap very quickly.

We'll keep groups as a requirement - but limit it to just one for now.

You say presence of groups enables the feature or setting a priority policy its functionally equivelant since a group is required.  I'll update the doc to say the policy activates it - but since both are required its functionally equiv